#161678177 : Add notifications model

gitau-kinuthia opened the pull request:
#### What does this PR do?
- Add Notifications Model
- configure django_q and django_cache


#### Description of Task to be completed?
- Add a notification model based on the nomenclature [here](https://www.w3.org/TR/notifications/)
- Add signals to create notifications to admins when a new incident report is filed
- Add signals to notify the user when the transition state for an incident report raised is changed


#### How should this be manually tested



#### Any background context you want to provide?
- To setup the Django cache which is required for  monitoring `django_q` ([details](https://docs.djangoproject.com/en/1.8/topics/cache/#database-caching)) run: 
```
pipenv run python manage.py createcachetable
```

- The nomenclature for the notifications is based on the spec in www.w3.org/TR/notifications/. 
The expected body of notifications is :
Generalized Format for notifications body: as proposed in the [django_notifications ](https://github.com/django-notifications/django-notifications/blob/master/notifications/base/models.py)package
 ```
        <actor> <verb> <time>

        <actor> <verb> <target> <time>

        <actor> <verb> <action_object> <target> <time>
```


#### What are the relevant pivotal tracker stories?
[161678177](https://www.pivotaltracker.com/story/show/161678177)

#### Postman Documentation


#### Screenshots (If applicable)

gitau-kinuthia committed:
ft-add-notifications: Add notifications model

-  Add notifications model
- add signals to create notifications
- add unit tests
- configure django_q and django_cache
- resolve merge conflicts

[Delivers #161678177]

gitau-kinuthia committed:
ft-add-notifications: Add notifications model

-  Add notifications model
- add signals to create notifications
- add unit tests
- configure django_q and django_cache
- resolve merge conflicts

[Delivers #161678177]

gitau-kinuthia committed:
ft-add-notifications: Add notifications model

-  Add notifications model
- add signals to create notifications
- add unit tests
- configure django_q and django_cache
- resolve merge conflicts
- update README.md

[Delivers #161678177]

kenneth051 commented on api/signals.py:
```diff
@@ -30,3 +42,61 @@ def track_application_actions(sender, instance, **kwargs):
+    instance = kwargs["instance"]
+    # if a new notification has been  created
+    if kwargs.get("created", False) is True:
+        notification_body = "{} submitted a new incident report {}."
```
@fgkinus good work, I have suggestions. the notification details could be added in the constant.py file. we would just call them from there then format them if we need to add contents to them. I mean the notification body, title. we could also do away with line 60 and just add the instance.submitted_by directly to origin on line 69.

gitau-kinuthia replied:
Thanks for these notes. I agree and  will incorporate the feedback 👍

kenneth051 commented on api/signals.py:
```diff
@@ -30,3 +42,61 @@ def track_application_actions(sender, instance, **kwargs):
+    """
+    instance = kwargs["instance"]
+    # notification details
+    incident_url = "{}/{}".format(reverse("incidence-reports-list"), instance.id)
```
we could also add these to the constant.py file

kenneth051 commented on core/models/notification.py:
```diff
@@ -0,0 +1,84 @@
+        <user> <closed> <issue 70> on <topic> <about 2 hours ago>
+    """
+
+    title = models.CharField(
```
I also see that the model fields specified in the story are not the same as you have here.. could you reach out to the FPC and have the story description updated.

gitau-kinuthia replied:
I will reach out and have the story updated 👍

kmazi replied:
Let's tailor the generalized format to our specific use case as much as possible. Nice job

kenneth051 commented on core/tests/test_asset_incident_report_model.py:
```diff
@@ -23,3 +28,48 @@ def test_model_string_representation(self):
+        self.assertEqual(notification.origin, report.submitted_by)
+        self.assertEqual(notification.title, "New Incident Report")
+
+        self.assertEqual(notification.body, body)
```
Is there a way to test for failure when a notification is not sent.

gitau-kinuthia replied:
I think to test when a notification not sent would entail counting notifications before triggering an event that would usually generate a notification and then verifying that notification count does not change.

kmazi replied:
Appears to me like this test would serve that purpose since it fails when notification is not sent. @kenneth051

kmazi commented on core/tests/test_asset_incident_report_model.py:
```diff
@@ -23,3 +28,48 @@ def test_model_string_representation(self):
         )
+
+
+class TestAssetIncidentReportNotifications(CoreBaseTestCase):
```
Nice categorization of test cases.

kmazi commented on core/tests/test_asset_incident_report_model.py:
```diff
@@ -23,3 +28,48 @@ def test_model_string_representation(self):
+
+class TestAssetIncidentReportNotifications(CoreBaseTestCase):
+    """
+    Test Notifications elicited by Asset Incident Report creation
```
Just for semantics sake and nothing related to tech, `elicited` isn't the right word to use there. looool

gitau-kinuthia replied:
Noted and fixed

kmazi commented on core/tests/test_asset_incident_report_model.py:
```diff
@@ -23,3 +28,48 @@ def test_model_string_representation(self):
+        notifications = self.Notifications_model.objects.filter(target__in=admins)
+        notification = notifications[0]
+        incident_url = "{}/{}".format(reverse("incidence-reports-list"), report.id)
+        body = "{} submitted a new incident report {}.".format(
```
Let's stick to [fstrings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings) for formatting. @kenneth051 what's your take on this?

kmazi commented on core/models/notification.py:
```diff
@@ -0,0 +1,84 @@
+
+class Notifications(models.Model):
+    """
+    A model representing the a notification object
```
slight typo

kmazi commented on core/tests/test_notifications_model.py:
```diff
@@ -0,0 +1,32 @@
+
+    def test_get_time_since(self):
+        timesince = self.notification.timesince()
+        self.assertTrue(isinstance(timesince, str))
```
This assertion doesn't show the time since notification was created but rather asserts the type.

gitau-kinuthia replied:
Any ideas on a test that would work. Knowing that the current time will be an unknown variable i.e I cannot think of a way to capture the actual time the test runs so as to do a value assertion

gitau-kinuthia replied:
> This assertion doesn't show the time since notification was created but rather asserts the type.
I figured a type assertion would make sure would make sure that a string representaion of time is actually returned veryfing the function works

kmazi commented on core/constants.py:
```diff
@@ -121,3 +120,14 @@
 
 TABLES = ["core_history", "django_admin_log"]
+
+NOTIFICATION_TITLES = ()
```
What's the essence of this constant?

gitau-kinuthia replied:
It was supposed to be a tuple containing valid `Notification Titles` for the notifications. title field. I decided against this approach for flexibility sake but didn't remove the variable. I have fixed this though

kmazi commented on api/signals.py:
```diff
@@ -30,3 +42,61 @@ def track_application_actions(sender, instance, **kwargs):
+        incident_url = "{}/{}".format(reverse("incidence-reports-list"), instance.id)
+
+        # create notifications to te different admins
+        for target in notification_targets:
```
This is a good place to use [atomic commit](https://docs.djangoproject.com/en/2.2/topics/db/transactions/#django.db.transaction.atomic) to ensure this is done with a single call to the database in a transaction.

gitau-kinuthia replied:
Thanks for this feedback. This actually something I had not considered  👍

kmazi commented on api/signals.py:
```diff
@@ -30,3 +42,61 @@ def track_application_actions(sender, instance, **kwargs):
+    instance = kwargs["instance"]
+    # notification details
+    incident_url = "{}/{}".format(reverse("incidence-reports-list"), instance.id)
+    notification_body = "The status of incident report {} that you filed has been updated to `{}`.".format(
```
`F-strings` is kinda cleaner in string literals interpolation.

gitau-kinuthia committed:
ft-add-notifications: Add notifications model

-  Add notifications model
- add signals to create notifications
- add unit tests
- configure django_q and django_cache
- resolve merge conflicts
- update README.md
- apply PR feedback

[Delivers #161678177]

kmazi commented on core/constants.py:
```diff
@@ -121,7 +122,16 @@
+INCIDENT_REPORT_STATUS_UPDATED_NOTIFICATION_TITLE = "Incident Report Status Update"
+# Notification Bodies
+INCIDENT_REPORT_CREATED_NOTIFICATION_BODY = Template(
+    "$email submitted a new incident report {$link}."
```
Are the curly braces intentional seeing that they would display in the resulting string?

gitau-kinuthia replied:
@kmazi Thanks for the noting the mistake. I have removed them

gitau-kinuthia committed:
ft-add-notifications: Add notifications model

-  Add notifications model
- add signals to create notifications
- add unit tests
- configure django_q and django_cache
- resolve merge conflicts
- update README.md
- apply PR feedback
- remove curly braces from notification body template

[Delivers #161678177]
